File: java/implement-trie-prefix-tree.java
class Trie {
    
    Trie[] next;
    boolean isEnd;

    /** Initialize your data structure here. */
    public Trie() {
        this.next = new Trie[26];
        this.isEnd = false;
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        Trie cur = this;
        for (char c: word.toCharArray()) {
            int i = c-'a';
            if (cur.next[i] == null) {
                cur.next[i] = new Trie();
            }
            cur = cur.next[i];
        }
        cur.isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        Trie cur = this;
        for (char c: word.toCharArray()) {
            int i = c-'a';
            if (cur.next[i] == null) return false;
            cur = cur.next[i];
        }
        return cur.isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        Trie cur = this;
        for (char c: prefix.toCharArray()) {
            int i = c-'a';
            if (cur.next[i] == null) return false;
            cur = cur.next[i];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: go/implement-trie-prefix-tree.go
type Trie struct {
	Next  map[rune]*Trie
	IsEnd bool
}

/** Initialize your data structure here. */
func Constructor() Trie {
	this := Trie{}
	this.Next = map[rune]*Trie{}
	this.IsEnd = false
	return this
}

/** Inserts a word into the trie. */
func (this *Trie) Insert(word string) {
	cur := this
	for _, c := range word {
		if cur.Next[c] == nil {
			newTrie := Constructor()
			cur.Next[c] = &newTrie
		}
		cur = cur.Next[c]
	}
	cur.IsEnd = true
}

/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
	cur := this
	for _, c := range word {
		if cur.Next[c] == nil {
			return false
		}
		cur = cur.Next[c]
	}
	return cur.IsEnd
}

/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
	cur := this
	for _, c := range prefix {
		if cur.Next[c] == nil {
			return false
		}
		cur = cur.Next[c]
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */